File: scm-csharp/scheme/primitives/PrimitiveJoinTimeoutExceptionP.cs
namespace scheme;

public class PrimitiveJoinTimeoutExceptionP : Primitive
{
    public override string Name() => "join-timeout-exception?";

    public override string Info() =>
        "Syntax: (join-timeout-exception? obj)\n" +
        "Library: (srfi 18)\n" +
        "Description: Returns #t if obj is a join-timeout exception.\n" +
        "Example:\n" +
        "  (guard (e ((join-timeout-exception? e) 'timeout)) (thread-join! t 0))";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        return IsThreadException(arguments[0], ThreadExceptionKind.JOIN_TIMEOUT) ? Value.T : Value.F;
    }

    internal static bool IsThreadException(object obj, ThreadExceptionKind kind)
    {
        if (obj is ErrorObject eo && eo.Irritants.Length > 0
            && eo.Irritants[0] is NativeValue nv && nv.value is SchemeThreadException te)
            return te.kind == kind;
        if (obj is NativeValue nv2 && nv2.value is SchemeThreadException te2)
            return te2.kind == kind;
        return false;
    }

    internal static SchemeThreadException? GetThreadException(object obj)
    {
        if (obj is ErrorObject eo && eo.Irritants.Length > 0
            && eo.Irritants[0] is NativeValue nv && nv.value is SchemeThreadException te)
            return te;
        if (obj is NativeValue nv2 && nv2.value is SchemeThreadException te2)
            return te2;
        return null;
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveJoinTimeoutExceptionP.java
package scheme.primitives;
import scheme.*;

public class PrimitiveJoinTimeoutExceptionP extends Primitive {
    @Override
    public String name() { return "join-timeout-exception?"; }

    @Override
    public String info() {
        return "Syntax: (join-timeout-exception? obj)\n" +
               "Library: (srfi 18)\n" +
               "Description: Returns #t if obj is a join-timeout exception.\n" +
               "Example:\n" +
               "  (guard (e ((join-timeout-exception? e) 'timeout)) (thread-join! t 0))";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);
        return isThreadException(arguments[0], SchemeThreadException.Kind.JOIN_TIMEOUT) ? Value.T : Value.F;
    }

    public static boolean isThreadException(Object obj, SchemeThreadException.Kind kind) {
        if (obj instanceof ErrorObject) {
            ErrorObject eo = (ErrorObject) obj;
            if (eo.irritants.length > 0 && eo.irritants[0] instanceof NativeValue) {
                NativeValue nv = (NativeValue) eo.irritants[0];
                if (nv.value instanceof SchemeThreadException)
                    return ((SchemeThreadException) nv.value).kind == kind;
            }
        }
        if (obj instanceof NativeValue) {
            NativeValue nv = (NativeValue) obj;
            if (nv.value instanceof SchemeThreadException)
                return ((SchemeThreadException) nv.value).kind == kind;
        }
        return false;
    }

    public static SchemeThreadException getThreadException(Object obj) {
        if (obj instanceof ErrorObject) {
            ErrorObject eo = (ErrorObject) obj;
            if (eo.irritants.length > 0 && eo.irritants[0] instanceof NativeValue) {
                NativeValue nv = (NativeValue) eo.irritants[0];
                if (nv.value instanceof SchemeThreadException)
                    return (SchemeThreadException) nv.value;
            }
        }
        if (obj instanceof NativeValue) {
            NativeValue nv = (NativeValue) obj;
            if (nv.value instanceof SchemeThreadException)
                return (SchemeThreadException) nv.value;
        }
        return null;
    }
}
